### DISCUSSION TITLE
SPA data fetching strategies

### DISCUSSION THREAD

**Discussion by localusercamp** [nuxt/nuxt#16734] [Questions]
## Introduction

**Hello!**
Before I ask a question, I would like to give a little context. I have been using **nuxt 2** in production for quite some time now (about three years). For now **nuxt 2** is huge ecosystem, which includes tons of great decisions made of you guys. Thank to all of you, it's a tremendous work, really. During this time, I have studied quite a few different practices adopted in the framework. 

One of them is how to _correctly_ fetch data from the backend. There are several ways to do this, but in my opinion the most flexible and easiest to configure is [nuxt/axios](https://github.com/nuxt-community/axios-module) module. It includes all the benefits of axios but also includes **request hooks**, some **useful helpers** and **context injection**.
Now, in modern web development we have standart pair of Backend-server and SPA-server (SRR or SSG). Backends contain more and more routing groups as they grow (especially if it's a microservice architecture). That's why i prefer to create different axios instances for different route groups and inject them, for example: 

`plugins/axios.js`
```javascript
...

inject('auth', axios.create({
  baseUrl: 'https://auth.domain.com', 
  withCredentails: true 
}));

inject('api', axios.create({
  baseUrl: 'https://api.domain.com/v1',
  headers: { 
    'Accept': 'application/json',
    'Content-Type': 'application/json',
  }
}));

...
```

`pages/example.vue`
```javascript
...

const { user } = await this.$auth.$post('login');
const { posts } = await this.$api.$get(`user/${user.id}/posts`);
...
```

It helps keep code clean and simple as you can see.

## Question: 📁 `server`

When i started to digging deeper in new **nuxt 3** feature called "server folder", I became less and less aware of its direct purpose.

- What are exact purposes for server folder (with some use cases)?
- How can i reuse my `useFetch` configuration as i showed on axios example?
- Why should i use `useFetch` instead of axios and what betefits it's brings in case of code cleanliness and less code repetition?

**I think answer for this question will help many people to better understand new great feature and its benefits. 📚**

**Thank you**

**Comment by sevillaarvin** [comment]
> * What are exact purposes for server folder (with some use cases)?

I'm not sure what answer you are looking for, from the docs:

"Nuxt automatically scans files inside the ~/server/api, ~/server/routes, and ~/server/middleware directories to register API and server handlers with HMR support."

https://v3.nuxtjs.org/guide/directory-structure/server#server-directory

If I have to put it in my words, server folder's purpose is a place to put your backend route handlers. If we remove everything Nuxt has and only have the `server/` folder, then it will function like `Express` node backend framework.

It is powered by `Nitro` and `h3` according to this: https://v3.nuxtjs.org/guide/concepts/server-engine/

----

> * How can i reuse my `useFetch` configuration as i showed on axios example?

One approach: https://github.com/nuxt/framework/issues/1852#issuecomment-966161569

> This is by design. `useFetch` shouldn't be triggered in a method; like most composables, it is designed to be run in `setup()`. If you need an on-demand fetch use `$fetch()` instead. Alternatively, you can implement like so:
> 
> ```js
> function useMyThing () {
>   const { data, refresh } = 
>   const [{ data: organization, refresh: refreshOrganization }, { data: repos, refresh: refreshRepos }] = Promise.all([
>     useFetch('https://api.github.com/orgs/nuxt'),
>     useFetch('https://api.github.com/orgs/nuxt/repos'),
>   ])
>   return {
>     organization,
>     repos,
>     fetch: () => Promise.all([refreshOrganization(), refreshRepos()])
>   }
> }
> ```

Another approach, this is what I did (note that I didn't know about composables back then, so this is what I naively came up with):

https://github.com/nuxt/framework/discussions/7719#discussioncomment-3709817

Maybe other people have different approaches, try searching the discussion.

----

> * Why should i use `useFetch` instead of axios and what betefits it's brings in case of code cleanliness and less code repetition?

`useFetch` is powered by [ohmyfetch](https://github.com/unjs/ohmyfetch/), which is just a minimal wrapper around the native [fetch.](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)

So if we compare `ohmyfetch` and `axios`, one benefit would be it is lightweight (didn't test, just my guess).

As for `useFetch`, it is just a wrapper around `useAsyncData`, which you can use (but not necessarily) as a wrapper for `ohmyfetch`

**useFetch**
> This composable provides a convenient wrapper around useAsyncData and $fetch. It automatically generates a key based on URL and fetch options, provides type hints for request url based on server routes, and infers API response type.

https://v3.nuxtjs.org/getting-started/data-fetching#usefetch

Ok, so you may ask, why would you use useFetch/useAsyncData vs axios? According to the docs, useFetch/useAsyncData has numerous helper `refs` and functions,  one of which is the ability to refresh/reexecute a request. If you use the lazy version, there will be a `pending` ref which you can use in your `<template>`.

https://v3.nuxtjs.org/getting-started/data-fetching#uselazyfetch

There are many more features/options and I'm still learning like you so I don't know them all.

Hope this helps

**Reply by sevillaarvin** [reply]
Found this discussion on Nuxt 3 and axios

https://github.com/nuxt/framework/discussions/4514

**Reply by localusercamp** [reply]
Thank you for your answer! I read more about `ohmyfetch` I found one cool thing with `$fetch.create()`. It will create a new $fetch instance with defaults like in axios.
